### ISSUE TITLE
java.langAssertionError Had an update, but it didn't match any rows!

### ISSUE THREAD

**Issue by poggenpower** [signalapp/Signal-Android#10726]
Every time I start signal, it crashes within a few seconds. 
If i switch on flight mode, the App stays alive and I can do any action the would work offline. 

Even preparing an answer seems to be possible and it gets somehow delivered before it crashes if I go online.

If I am offline I can also check the debug log, but submit doesn't work as of missing internet connection. Unfortunately Copy and Past doesn't work from the Debug log either, therefor here a retyped error message:
```
SignalUncoughtExceptionHandler: java.langAssertionError Had an update, but it didn't match any rows!
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.database.RecipientDatabase.applyStorageUpdate(RecipientDatabase.java:897)
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.jobs.StorageSyncJob.performSync(StorageSyncJob.java:201)
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.jobs.StorageSyncJob.onRun(StorageSyncJob.java:108)
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.jobs.BaseJob.run(BaseJob.java:30)
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:86)
SignalUncoughtExceptionHandler:  at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:49)
```
Signal Version 5.2.3 (772, 77202)

This crash behaviour started at the same time as my contacts complained about "bad encrypted messages" sended by me. 
Another related thing could be the activation on Signal on an iPad, that was not active for around 4 weeks and linked with my account. 

Signal crashes on may Android Phone only. Linked desktop Client and iPad App is working fine.

**Comment by bierbo** [comment]
The same error occurs on my phone since yesterday.
The error message came out of nowhere and since then at very short intervals.
If I'm quick, I can load the app and check 1-2 things, then the error message appears again. A force kill did not help, after restarting the app I am stuck in the error message loop again.

**Comment by lachlansleight** [comment]
This is also happening to me - the app loads new messages, which appear in my notifications and then immediately crashes with the following adb logcat output:

```
01-27 22:32:05.617 17817 17888 E AndroidRuntime: FATAL EXCEPTION: signal-JobRunner-1
01-27 22:32:05.617 17817 17888 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 17817
01-27 22:32:05.617 17817 17888 E AndroidRuntime: java.lang.AssertionError: Had an update, but it didn't match any rows!
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.database.RecipientDatabase.applyStorageSyncUpdates(RecipientDatabase.java:873)
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.jobs.StorageSyncJob.performSync(StorageSyncJob.java:203)
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.jobs.StorageSyncJob.onRun(StorageSyncJob.java:110)
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.jobs.BaseJob.run(BaseJob.java:25)
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:86)
01-27 22:32:05.617 17817 17888 E AndroidRuntime:        at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:49)
01-27 22:32:05.613 17817 17888 I chatty  : uid=10228(org.thoughtcrime.securesms) signal-JobRunne identical 3 lines
01-27 22:32:05.614 17817 17888 V Cursor  : Filling cursor window with start position:0 required position:0
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler: java.lang.AssertionError: Had an update, but it didn't match any rows!
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.database.RecipientDatabase.applyStorageSyncUpdates(RecipientDatabase.java:873)
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.jobs.StorageSyncJob.performSync(StorageSyncJob.java:203)
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.jobs.StorageSyncJob.onRun(StorageSyncJob.java:110)
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.jobs.BaseJob.run(BaseJob.java:25)
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:86)
01-27 22:32:05.618 17817 17888 E SignalUncaughtExceptionHandler:        at org.thoughtcrime.securesms.jobmanager.JobRunner.run(JobRunner.java:49)
```

No improvement upon restarting the device, killing the app, etc. (This is on a Xiaomi 9T Pro). I was running 5.0.something, just updated to 5.3.6 and am getting the same problem. The app was fine this morning.

This is actually a huge problem for me - Signal is my primary messaging application, I haven't used SMS in years! Without Signal I'm effectively cut off from all my friends and family, at least on my phone. The PC app still works...

**Comment by hiqua** [comment]
If you can make a backup, first make it and keep it somewhere so that you can minimize the losses. The apk for a previous version is still available:

https://updates.signal.org/android/Signal-Android-website-prod-universal-release-5.1.9.apk

I would try to restore the backup with this version and see if it works, then block updates for the time being, hoping that this bug is solved.

From what you said and the logs, it might be related to this commit: https://github.com/signalapp/Signal-Android/commit/a517fc4e15df86cce89b808cbde8a4cdcc913bad

**Comment by lachlansleight** [comment]
I was actually having this same problem on 5.0.x (I don't remember exactly what the version was) - I updated it hoping that it would fix the issue but it didn't. So I don't think downgrading to 5.1.9 will help, but I'm happy to try it if you think it'll help (I'd like confirmation first since I won't want to risk losing data for something that isn't going to yield any better information)

**Comment by hiqua** [comment]
OK sorry I missed the "5.0", them my advice doesn't apply.

**Comment by lachlansleight** [comment]
This morning the problem mysteriously stopped happening. My app is now version 5.3.7 so I suspect that's involved?

**Comment by poggenpower** [comment]
I have installed 5.3.9 and still this issue. One of feeled 1000 starts it keeps up even if connected to the internet. As already said. If I put my phone into fight mode, signal keeps running. But if Network connection is back it crashes typically immediately.

**Comment by lachlansleight** [comment]
Yes my app has broken again now, was working this morning and is now crashing immediately again.

So I'm now basically unable to contact anyone I know when I'm outside the house.

How ironic, I finally succeeded in my multi-year campaign to get everyone I know on signal, and now I'm screwed because of it lol. Hopefully this bug can be hunted down soon because switching away from signal is no longer a realistic option for me :'(

**Comment by hiqua** [comment]
Possible ways to troubleshoot this further (I don't have this bug and I don't know how to reproduce it so these are just guesses):
* if one of you doesn't mind losing their old messages, you could try to reinstall Signal and start afresh. Could it be some kind of corruption of the local db? Given that your other devices seem to work fine, I would say that's the most likely explanation.
  * you could also try to backup, delete, restore backup, but given this problem it's possible that the restoration doesn't work

* if the app crashes when connected, then doesn't when in airplane mode, can you submit a log after you've reopened it? Maybe it still contains some info? Maybe there are some errors also when there is no apparent crash?

**Comment by hiqua** [comment]
> I have installed 5.3.9 and still this issue.

Did you get this version from Google Play? It seems that the latest version is 5.3.7.1. It could be just a regional thing, but given that 5.3.7.1 has a commit that 5.3.9 doesn't have, I would find it surprising then that the latest version available is not 5.3.10 instead.

The commit is this one: https://github.com/signalapp/Signal-Android/commit/3f51f89d866081d91ee576eea76af1df39b0736f

Not sure why this warranted a 5.3.7.1 instead of just waiting for the next 5.3.n release.

**Comment by poggenpower** [comment]
@hiqua I am subscribed as beta tester via PlayStore, therefore - I think - I get more recent releases.

**Comment by poggenpower** [comment]
Unfortunately I can't get the log. If I am online it crashes to fast, if i am offline I cant send the logs. It is also not possible to copy the log or savd it locally to send it out of Band.
I would really appreciate if i can keep my messages. 
Some further details: I am still able to receive messages, even with attachments, but I need to be offline to read them. Similar for sending messages. It is possible to write messages while offline. Even if signal is ceashing if I switch to online the message will bf send. As this issue occured a whlle ago I have received and send messages in most of my groups and contacts.

**Comment by hiqua** [comment]
Are the others using beta versions as well? That could have been a bug in beta that never happened in the stable releases, which could explain why so "few" people seem to have it.

**Comment by poggenpower** [comment]
Does it help to share the crash report Android is generating?

**Comment by poggenpower** [comment]
Just found out that I am already at 5.3.10 there was an update yesterday.

**Comment by hiqua** [comment]
> Does it help to share the crash report Android is generating?

It cannot hurt, but you should double-check what's in it, it could contain some private information.

**Comment by lachlansleight** [comment]
The crashreport that I get from Android directly is just the adb output that I shared above - I'm happy to get more logs but would need instructions on how to go about getting them. I tried `adb bugreport` but I don't think that's super relevant to this crash

**Comment by poggenpower** [comment]
I have just reset my complete Smartphone and reinstalled Signal and restored from Backup. But it is behaving the same. I found out that I can't copy/past from the Android crash report either. Signal was the 3rd App I had installed.

**Comment by hiqua** [comment]
@lachlansleight are you also on the beta channel, or the stable one?

I think the only next step you guys could take is debug the app with Android Studio and have a breakpoint exactly where it fails, and check what exactly in the db update fails (table / column). Obviously that's way more involved...

**Comment by lachlansleight** [comment]
@hiqua I'm assuming the stable one - I just downloaded Signal from the Play store normally.

How exactly would I go about that? I just cloned the project and opened it up in Android Studio, but I usually work in Unity so I have no idea how to debug the app / set breakpoints, etc 😅

Is the idea that I can connect Android Studio to my device and watch for breakpoints to be hit when I run the app, or would I be building from source and running it in some kind of debug mode on my device?

**Comment by hiqua** [comment]
I think the easiest would be to backup from your phone, save your backup to your desktop, then start the Signal apk from an emulator (from Android Studio) and restore the backup there. The problem is that the emulator will then be the primary device and you won't have access on your phone anymore.

Not 100% sure, but I don't think it's possible to debug the normal release since it's probably not a debug release. It is possible to install a debug release on your device and debug it there, but it will probably be a bit more complicated. Note that you won't be able to install your own build over the one from the Signal team (because the certificate won't match).

To be clear, it's a long shot. But if you have the time / motivation / skills, why not.

P.S.: to be clearer, I'm not affiliated with the Signal team, I'm just concerned that there might be such a bug in Signal. There's no guarantee that they'd fix the bug even if we find it. But I think that if we can pinpoint the issue to a specific line they might fix it eventually.

**Comment by poggenpower** [comment]
Since I have reset Signal the links to my Desktop Apps are lost, but I can't recreate as they need a network connection and if Signal is connected to the network it is crashing.

**Comment by poggenpower** [comment]
I was able to upload the debug log https://debuglogs.org/3e22a6606b1714b2ff359e5b63edb947bfaffae57d39c64aff989adbfb9544ea as the app stayed open for a while unexpectedly.

**Comment by poggenpower** [comment]
@lachlansleight the beta program is indicated by a note right below the update & open button in the playstore.

**Comment by lachlansleight** [comment]
I think doing a full Android Studio debug is probably beyond me right now - I'm pretty busy with work so don't really have the time to devote to another open source project hehe...

That debug log is interesting poggenpower, both the 'failed to insert' repetitive error, and the final exception at the bottom. 

I hope someone from the Signal team sees this thread! I really am kind of freaking out at the fact that I can't use signal on my phone - it's my main form of communication!

**Comment by poggenpower** [comment]
After I was able to submit the debug logs signal was crashing again. But from the logs it looks like it have problems to sync users. I did several things trigger a clean up. I have deleted all "Jane Doe is now.on signal" conversations. Have deleted not archived them. Have rejected the permission to access contacts. Left some orphaned groups. Grant access to contacts again. While doing all these steps signal kept longer alive before crashing. Now it is running for more than 20 minutes without a crash

**Comment by hiqua** [comment]
Thanks for the log @poggenpower ! There are some errors regarding group migration, I'd suggest you leave groups you don't use anymore, if you have some. Do you still have V1 groups? If you manage to get it running again, maybe try to play with the settings that are stored on the servers, like deactivate the pin setting, reactivate it with a new one... Again this is a long shot.

@lachlansleight can you check at the beginning of the log if you see similarities to your situation? E.g.:
* Android 11
* pin activated

**Comment by lachlansleight** [comment]
@hiqua I'm on Android 10, and my pin state in the log is identical except `PIN Required at Registration` is `false`

Today I tried deleting all my old conversation threads, archived or otherwise (some 500+ threads!) and that didn't resolve the problem.

I also tried deleting all conversations that involved groups (some of which were legacy groups), and that also didn't help.

**Comment by poggenpower** [comment]
Had another crash this morning while touching an old group. Unfortunately I have not recorded the group version. An signal kept crashing until I have completely deleted (not only archived) this group. Here the log https://debuglogs.org/4d55c2dc3e55a05fa3bc8ae73955e98cd9a077418ec7c861c978ea5902d26418

**Comment by poggenpower** [comment]
The crashing pattern has slightly changed now. If Signal is crashing. It seems enough to go to flight mode to start signal. If signal is running I can go online again and signal keeps running.
@hiqua the commits above are promising

**Comment by hiqua** [comment]
> I also tried deleting all conversations that involved groups (some of which were legacy groups), and that also didn't help.

I'm not sure deleting the conversation could do much, leaving the group would be more interesting.

But let's wait for 5.3.11 and maybe things will be solved there.

**Comment by poggenpower** [comment]
Have now 5.3.11 installed and some groups appeared again that I have not seen before anymore but still member. All of them were older groups created before the new group features. 
But it looks like all groups are migrated to the new format now. 
Signal seems to be stable so far. Looks like the fixes mention by @hiqua really helped with this issue.

**Comment by hiqua** [comment]
@poggenpower while it's stable can you submit a debug log just in case the bug is still not completely solved?

**Comment by lachlansleight** [comment]
5.3.12 seems to have fixed things for me - will monitor over the next couple of days

I submitted a debug log: https://debuglogs.org/69813989d2d8166796798f8808ff3cf82b6166b9b1e0171511dcfddec47b29d5

**Comment by hiqua** [comment]
I think the last error was on 5.3.10, no error since 5.3.12

**Comment by greyson-signal** [comment]
Hey yes, there's a fix for this in 5.3.12. For the curious, the underlying problem was that in rare situations, we were updating the same contact twice, and the second update would fail. That should hopefully be fixed now!

**Comment by hiqua** [comment]
@greyson-signal awesome, thanks for the fix and the info!
